### ISSUE TITLE
[contracts] Add isStateTerminal check in MixinSetState

### ISSUE THREAD

**Issue by snario** [counterfactual/monorepo#1310]
Right now only `MixinSetStateWithAction` checks for the `isStateTerminal` condition. We can add it to `MixinSetState` as well. This is useful in the case of a basic unanimous consent payment channels app for example.

**Comment by adklempner** [comment]
Am I correct that this will require including the entire `bytes appState` in `SignedAppChallengeUpdate` and not just the `bytes32 appStateHash`?

**Comment by AlexXiong97** [comment]
@snario Currently, I believe you could still do this without any change to contracts very implicitly by setting `req.timeout` to `0` --> assumption is Node will check `isStateTerminal` before countersign on a setState whose timeout field is 0.

**Comment by snario** [comment]
@AlexXiong97 Yes that's right actually; we can set `timeout` to `0`. Basically by doing that it is an implicit "this state is terminal regardless of what the contract says" statement by the users of the app.

I think this issue, however, is specifically describing the scenario where for example in a game of TTT both parties sign the state that "Bob has won with 3 in a row" and _then_ the counterparty goes offline (before signing the `UninstallCommitment` via the uninstall protocol). In that case specifically, you will want to submit this latest state to the `setState` function with the `isStateTerminal` flag.

We could just avoid this though by making it a Node _requirement_ (part of the implicit set of state channels convention we're introducing with this project) that when a state is terminal, you always set `timeout` to `0` like @AlexXiong97 suggested.

**Comment by AlexXiong97** [comment]
Indeed. Personally, I don't know which would be a better engineering choice.

Keeping it a requirement to Node seems like a "leaky abstraction" which may complicate debugging and may violate some modularity principle, meaning future contributor will need to know this decision that binds the two separate modules.

While requiring `appState` and `terminalCheck` flag in the `setState` request increases communication load/overhead (although negligible in few party channel, but can get expensive when appState is huge for complicated channelized app and for large n-party channel)

Maybe @IIIIllllIIIIllllIIIIllllIIIIllllIIIIll can give some thoughts as well
